server/s3_data_usage.cc

```diff
@@ -0,0 +1,511 @@
+  s3_log(S3_LOG_DEBUG, "", "%s Entry\n", __func__);
+  if (!singleton) {
+    s3_log(S3_LOG_INFO, "", "%s Create a new DataUsageCache", __func__);
+    singleton.reset(new S3DataUsageCache());
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit\n", __func__);
+  return singleton.get();
+}
+
+void S3DataUsageCache::set_max_cache_size(size_t max_size) {
+  s3_log(S3_LOG_DEBUG, "", "%s Entry\n", __func__);
+  s3_log(S3_LOG_INFO, "", "Set data usage cache size to %zu", max_size);
+  max_cache_size = max_size;
+  s3_log(S3_LOG_DEBUG, "", "%s Exit\n", __func__);
+}
+
+std::string S3DataUsageCache::generate_cache_key(
+    std::shared_ptr<RequestObject> req) {
+  const std::string &req_id = req->get_stripped_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry\n", __func__);
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit\n", __func__);
+  return req->get_account_name();
+}
+
+std::string get_server_id() {
+  s3_log(S3_LOG_DEBUG, "", "%s Entry\n", __func__);
+  s3_log(S3_LOG_DEBUG, "", "%s Exit\n", __func__);
+  return "1";
+}
+
+std::shared_ptr<DataUsageItem> S3DataUsageCache::get_item(
+    std::shared_ptr<RequestObject> req) {
+  const std::string &req_id = req->get_stripped_request_id();
+  s3_log(S3_LOG_DEBUG, req_id, "%s Entry", __func__);
+
+  std::string key_in_cache = generate_cache_key(req);
+  const bool f_new = (items.end() == items.find(key_in_cache));
+
+  if (items.size() + f_new > max_cache_size) {
+    s3_log(S3_LOG_WARN, req_id, "Data usage cache is full");
+    if (!shrink(req_id)) {
+      s3_log(S3_LOG_WARN, req_id, "Failed to shrink the data usage cache");
+      return nullptr;
+    }
+  }
+
+  if (!f_new) {
+    s3_log(S3_LOG_INFO, req_id, "Found entry in Cache for %s",
+           key_in_cache.c_str());
+    s3_log(S3_LOG_DEBUG, req_id, "%s Exit", __func__);
+    return items[key_in_cache];
+  }
+
+  s3_log(S3_LOG_INFO, req_id, "Entry not found in Cache for %s",
+         key_in_cache.c_str());
+  auto subscriber = std::bind(&S3DataUsageCache::item_state_changed, this,
+                              std::placeholders::_1, std::placeholders::_2,
+                              std::placeholders::_3);
+  std::shared_ptr<DataUsageItem> item(
+      new DataUsageItem(req, key_in_cache, subscriber));
+  items.emplace(key_in_cache, std::move(item));
+
+  s3_log(S3_LOG_DEBUG, req_id, "%s Exit", __func__);
+  return items[key_in_cache];
+}
+
+bool S3DataUsageCache::shrink(const std::string &request_id) {
+  s3_log(S3_LOG_DEBUG, request_id, "%s Entry", __func__);
+  if (inactive_items.size() == 0) {
+    s3_log(S3_LOG_INFO, request_id,
+           "%s Cannot shrink cache: all items are active", __func__);
+    s3_log(S3_LOG_DEBUG, request_id, "%s Exit\n", __func__);
+    return false;
+  }
+
+  // The front item of the list is the LRU one.
+  DataUsageItem *item = inactive_items.front();
+  const std::string cache_key = item->cache_key;
+
+  s3_log(S3_LOG_DEBUG, request_id,
+         "Data usage for \"%s\" will be removed from the cache as inactive",
```

Comment:
1) LOG_INFO
2) `"Data usage cache entry for ...`

Reply:
Done.